### PR TITLE
refine OpKernelType

### LINKED ISSUE #6769: need to refine our design of OpKernelType
An operator can have different kernel implementations. Fluid uses `OpKernelType` to identify a unique Kernel. Currently, `OpKernelType` is defined as follows:
https://github.com/PaddlePaddle/Paddle/blob/2d5ec16bc8a09fb8e0f62c89b116b0cd1d333907/paddle/framework/operator.h#L348-L374

It contains two keys, `Place` and `DataType`. However, these two keys are not enough. We need a more complete representation of `OpKernelType`. There are four keys to determine a kernel type of an operator: `Place`/`Library`/`DataType`/`Layout`.

We often implement a kernel of an operator with some computing library in certain device. Please remind that computing library and device are not one-to-one corresponding. A device can have a lot of computing libraries and a computing library can also support several devices. 

For example, Eigen library can support Nvidia GPU/AMD GPU/CPU. And MKLDNN library can support Intel CPU/Intel FPGA. So, both `Place` and `Library` should be a key of `OpKernelType`.

Another two keys are `DataType` and `Layout`. It's obvious that different DataTypes, like fp64/fp32/int8 will have different kernels. And please refer to #6765 to get the details about `Layout`.

### PR DESCRIPTION
**Pull request by QiJune**
Fix #6769

### CODE REVIEW

File: paddle/framework/op_kernel_type.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+struct OpKernelType {
+  struct Hash {
+    size_t operator()(const OpKernelType& key) const {
+      int place = key.place_.which();
+      int data_type = static_cast<int>(key.data_type_);
+      int data_layout = static_cast<int>(key.data_layout_);
+      int library_type = static_cast<int>(key.library_type_);
+
+      size_t seed = 0;
+      hash_combine(seed, place);
```
Comment (dzhwinter) [inline paddle/framework/op_kernel_type.h]: how about combine these four fields as one integer, then make a hash_combine function call?

File: paddle/framework/op_kernel_type.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace paddle {
+namespace framework {
+
+/*
+Refer to https://stackoverflow.com/questions/35985960/
+c-why-is-boosthash-combine-the-best-way-to-combine-hash-values
+*/
+template <class T>
+inline void hash_combine(std::size_t& seed, const T& v) {
```
Comment (dzhwinter) [inline paddle/framework/op_kernel_type.h]: Google Style.

File: paddle/framework/op_kernel_type.h (hunk 1 above)
Comment (QiJune) [reply paddle/framework/op_kernel_type.h]: We can not simply combine these four fields together. HashCombine is the function to do such work.

File: paddle/framework/op_kernel_type.h (hunk 2 above)
Comment (QiJune) [reply paddle/framework/op_kernel_type.h]: Done

File: paddle/framework/op_kernel_type.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+namespace paddle {
+namespace framework {
+
+/*
+Refer to https://stackoverflow.com/questions/35985960/
+c-why-is-boosthash-combine-the-best-way-to-combine-hash-values
+*/
+template <class T>
+inline void HashCombine(std::size_t& seed, const T& v) {
```
Comment (jacquesqiao) [inline paddle/framework/op_kernel_type.h]: https://google.github.io/styleguide/cppguide.html#Function_Parameter_Ordering

Comment (QiJune) [reply paddle/framework/op_kernel_type.h]: Done

File: paddle/framework/op_kernel_type.h (hunk 1 above)
Comment (dzhwinter) [reply paddle/framework/op_kernel_type.h]: 我的意思是，四次HashCombine调用开销可以省略。

KernelType种类并不多，不会发生碰撞，hash只是让分布均匀一下。
防止碰撞可以这样
```c++
constexpr int SHIFT = 8; // every kind of type less than 2^8
int data_type = static_cast<int>(key.data_type_) + 1<<(SHIFT);
int data_layout = static_cast<int>(key.data_layout_) + 1 <<(SHIFT + 1);
int library_type = static_cast<int>(key.library_type_) + 1 << (SHIFT + 2);
int kernel_id = data_type + data_layout + library_type + place.which();
std::hash<int> hasher;
return hasher(kernel_id);
```

**Pull request merged**